File: web/src/features/addressBook.ts
/**
 * Address book (organisation-wide directory) frontend types.
 *
 * Mirrors app/models/address_book.py and documents/api-contract.md
 * "Organisation-wide address book" — see web/README.md for the screens
 * that consume these shapes.
 */

export interface GroupPathEntry {
  id: string;
  name: string;
}

export interface AddressBookGroupRole {
  kind: "group_role";
  role_type_name: string;
  group_id: string;
  group_name: string;
  group_path: GroupPathEntry[];
}

export interface AddressBookAccessRole {
  kind: "access_role";
  access_role_name: string;
}

export type AddressBookRole = AddressBookGroupRole | AddressBookAccessRole;

export interface AddressBookEntry {
  contact_id: string;
  first_name: string;
  last_name: string;
  email: string | null;
  roles: AddressBookRole[];
}

export interface VisibilityGroupRole {
  id: string;
  kind: "group_role";
  role_type_name: string;
  group_name: string;
  hidden_from_directory: boolean;
}

export interface VisibilityAccessRole {
  id: string;
  kind: "access_role";
  access_role_name: string;
  hidden_from_directory: boolean;
}

export type VisibilityRole = VisibilityGroupRole | VisibilityAccessRole;

export interface DirectoryVisibility {
  hidden_from_directory: boolean;
  roles: VisibilityRole[];
}

export interface DirectoryVisibilityPatch {
  hidden_from_directory?: boolean;
  roles?: Array<{ id: string; hidden_from_directory: boolean }>;
}

File: app/models/address_book.py
"""Address book (organisation-wide directory) API models.

See documents/api-contract.md "Organisation-wide address book" and
documents/family-and-directory-design.md for the rules these shapes encode.
"""

from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class GroupPathEntry(BaseModel):
    id: UUID
    name: str


class AddressBookGroupRole(BaseModel):
    kind: Literal["group_role"] = "group_role"
    role_type_name: str
    group_id: UUID
    group_name: str
    group_path: list[GroupPathEntry]


class AddressBookAccessRole(BaseModel):
    kind: Literal["access_role"] = "access_role"
    access_role_name: str


AddressBookRole = Annotated[
    AddressBookGroupRole | AddressBookAccessRole, Field(discriminator="kind")
]


class AddressBookEntry(BaseModel):
    contact_id: UUID
    first_name: str
    last_name: str
    email: str | None
    roles: list[AddressBookRole]


class AddressBookPage(BaseModel):
    items: list[AddressBookEntry]
    page: int
    page_size: int
    total: int


class VisibilityGroupRole(BaseModel):
    id: UUID
    kind: Literal["group_role"] = "group_role"
    role_type_name: str
    group_name: str
    hidden_from_directory: bool


class VisibilityAccessRole(BaseModel):
    id: UUID
    kind: Literal["access_role"] = "access_role"
    access_role_name: str
    hidden_from_directory: bool


VisibilityRole = Annotated[
    VisibilityGroupRole | VisibilityAccessRole, Field(discriminator="kind")
]


class DirectoryVisibility(BaseModel):
    hidden_from_directory: bool
    roles: list[VisibilityRole]


class RoleVisibilityPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    hidden_from_directory: bool


class DirectoryVisibilityPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # Defaults permit omission but the nonnullable annotations reject explicit
    # null, matching the ContactPatch convention (models/contacts.py).
    hidden_from_directory: bool = Field(default_factory=lambda: None)
    roles: list[RoleVisibilityPatch] = Field(default_factory=lambda: None)

    @model_validator(mode="after")
    def require_changes(self) -> "DirectoryVisibilityPatch":
        if not self.model_fields_set:
            raise ValueError("At least one of hidden_from_directory or roles is required")
        return self
